android_vs_ios.md

```diff
@@ -0,0 +1,175 @@
+
+## Common features and similarities (Kotlin and Swift)
+* Mainly object oriented (classes, inheritance, interface/protocol)
+* Statically typed
+* Type inference with option to specify explicit type
+* Type alias
+* Extensions
+* Generics
+* Optionals
+* Properties
+* Higher order functions
+* Default method parameters
+* Operator overloading
+* Interoperable with the previous language (with Java and Objective C respectively)
+* Semicolons are optional
+
+## Memory management
+Kotlin and Java uses [garbage collection](https://en.wikipedia.org/wiki/Garbage_collection_(computer_science)), which handles pretty much all the memory deallocation automatically, with the downside being that it can cause performance issues (ex: frame drop) when it runs and it is very hard/impossible to predict when it is triggered. Memory leaks can still happen (especially on Android), but there's not much concern for cyclic references.
+
+Swift and Objective C uses [automatic reference counting (ARC)](https://docs.swift.org/swift-book/LanguageGuide/AutomaticReferenceCounting.html), which makes it more predictable on when memory is deallocated, but requires more care to avoid retain cycles (ie: memory leaks). This happens when an object A references object B, and B also references back to A (ex: in MVP, View references Presenter while Presenter also references View, which is probably part of reason iOS prefers MVVM instead). This is solved by marking one of the reference as `weak` or `unowned` to break the cycle. This problem can also manifest itself when a lambda/closure references `self` inside of it and is held by a member variable (it’s not a problem if the lambda/closure is held by a local variable). This can be solved by using a [capture list](https://docs.swift.org/swift-book/ReferenceManual/Expressions.html#Capture%20Lists) and marking `self` as `weak` or `unowned`.
+
+## Struct vs Data Class
+Kotlin and Swift both offer another construct for representing objects in additional regular classes. Swift has [structs](https://docs.swift.org/swift-book/LanguageGuide/ClassesAndStructures.html#ID88) (which is a bit different from C structs), and Kotlin has [data classes](https://kotlinlang.org/docs/reference/data-classes.html). Swift's struct is not equivalent to Kotlin’s data class, although they sometimes are used for the same purpose.
+
+Data class’ main feature is to provide a default `equals()`, `hashcode()`, and `copy()` method and [destructuring](https://kotlinlang.org/docs/reference/multi-declarations.html). So it is still a class, but with a couple of functions already implemented.
+
+Struct’s main feature to is to serve as value type and enforce immutability, so it behaves differently from class, particularity when the object is passed around. (Swift 4.1 also added automatic synthesizer to implement Equatable and Hashable)
+
+They both forbid subclassing and are often be used for models.
+
+## Extensions + Protocol/Interface
```

Comment:
Also, is there any difference in what's accessible to extensions? For example, since Kotlin extensions resolve to static methods in Java, they can only access public members of a class. I don't know whether or not something similar holds for Swift.

Reply:
swift extensions also can't access private variables.